Use GenerateExportHeader

ClausKlein opened the pull request:
3e02e38 Set SPGLIB_SHARED_LIBS true for windows CI
454b448 Finalize under OSX
1cd7be3 Fix build problems under windows
8ccb1c9 Use macos-14 and macos-14 on CI
925448f Always include GNUInstallDirs cmake module
1e6d0ce Use GenerateExportHeader

ClausKlein committed:
Use GenerateExportHeader

ClausKlein committed:
Always include GNUInstallDirs cmake module

ClausKlein committed:
Use macos-14 and macos-14 on CI

ClausKlein committed:
Fix build problems under windows

LecrisUT commented:
Note: disable the `SPGLIB_SHARED_LIBS` in the `windows-ci` preset.

ClausKlein commented:
Why should I do it?

LecrisUT commented:
Because originally the tests only run for static-library which do not care about `dll_export`. This change would specifically affect the case of shared library which not being tested currently, and in https://github.com/spglib/spglib/pull/437 there are more relevant issues to arising from that.

ClausKlein committed:
Finalize under OSX

set CMAKE_C_VISIBILITY_PRESET too
set and install PUBLIC FILE_SET HEADERS
add the PRIVATE headers to library sources

ClausKlein committed:
Set SPGLIB_SHARED_LIBS true for windows CI

ClausKlein commented on src/niggli.h:
```diff
@@ -35,14 +35,17 @@
 int niggli_get_micro_version(void);
-int niggli_reduce(double* lattice_, const double eps_,
-                  const int aperiodic_axis);
+API_SPGLIB int niggli_reduce(double* lattice_, const double eps_,
+                             const int aperiodic_axis);
+API_SPGLIB void spg_set_error_code(SpglibError err);
```
Note: Non PUBLIC API used from tests, need to prevent unresolved symbols!

LecrisUT commented on test/CMakeLists.txt:
```diff
@@ -276,5 +282,12 @@ endfunction()
+# package/pkg-config/src -> ../src/
+# package/FetchContent/src -> ../src/
+# package/find_package/src -> ../src/
+if (UNIX AND NOT PROJECT_IS_TOP_LEVEL)
```
Please don't disable these. Even with symlinks they still run as shown in the other PR

ClausKlein replied:
That does not work under windows, the symlinks and the installed export package is not found in that way you build this tests!

But on UNIX, it still works: 
```
      Start 40: test-FetchContent
40/42 Test #40: test-FetchContent ........................................................   Passed    4.60 sec
      Start 41: test-find_package
41/42 Test #41: test-find_package ........................................................   Passed    2.07 sec
      Start 42: example-c_api
42/42 Test #42: example-c_api ............................................................   Passed    1.89 sec

100% tests passed, 0 tests failed out of 42
```

ClausKlein replied:
On linux, I install the package to a stage directory.
Then I cd to `test` and start the `workflow preset`.

On Windows, I dow the same in Visual Studio. only the install path is strange here?

LecrisUT replied:
Argh, I commented on the wrong PR, move this discussion to the main one

ClausKlein committed:
Build example subdir alway on CI again

Add .cmake-format and pip requirements.txt file too

ClausKlein committed:
Use ENVIRONMENT_MODIFICATION with cmake -E env

LecrisUT commented on test/CMakeLists.txt:
```diff
@@ -272,7 +272,10 @@ function(Spglib_add_test test)
 
 	add_test(NAME ${ARGS_TEST_NAME}
-			COMMAND ${CMAKE_CTEST_COMMAND} --build-and-test ${CMAKE_CURRENT_SOURCE_DIR}/${test}
+			COMMAND ${CMAKE_COMMAND} -E env --modify PATH=path_list_prepend:${CMAKE_INSTALL_PREFIX}/lib
```
This should be in test property to make debugging more legible.

But for the usage in `ctest --build-and-test` I completely disagree with adding it. The example and package tests need to be able to test as standalone builds, like the user would have run it, and populating and `ENV` hides issues in thr CMakeLists.txt file that the user should know beforehand.

Note that `testing-farm` tests the case of standalone test project with system installed version. It does not cover Windows, but at the same time window does not have a packaging system we can upload to.

LecrisUT replied:
Argh I commented on wrong PR again. I will unsubscribe here so I don't have both in notification

ClausKlein committed:
Disabe build of example_full.c on windows for now

ToDo: First fix compile errors under windows!

ClausKlein committed:
Default Preset works now

Use CMAKE_RUNTIME_OUTPUT_DIRECTORY as CWD while test

ClausKlein committed:
Windows CI test with DLL but without python for now!

ClausKlein committed:
Add cmake/SpglibGlobalSetup.cmake

add .envrc too
